streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java

```diff
@@ -907,62 +378,29 @@ public synchronized StreamsBuilder addStateStore(final StateStoreSupplier suppli
      * This {@link ProcessorNode} should be used to keep the {@link StateStore} up-to-date.
      * The default {@link TimestampExtractor} as specified in the {@link StreamsConfig config} is used.
      *
-     * @param storeSupplier         user defined state store supplier
+     * @param storeBuilder          user defined {@link StoreBuilder}; can't be {@code null}
      * @param sourceName            name of the {@link SourceNode} that will be automatically added
-     * @param keyDeserializer       the {@link Deserializer} to deserialize keys with
-     * @param valueDeserializer     the {@link Deserializer} to deserialize values with
      * @param topic                 the topic to source the data from
+     * @param consumed              the instance of {@link Consumed} used to define optional parameters; can't be {@code null}
      * @param processorName         the name of the {@link ProcessorSupplier}
      * @param stateUpdateSupplier   the instance of {@link ProcessorSupplier}
      * @return itself
      * @throws TopologyException if the processor of state is already registered
      */
-    public synchronized StreamsBuilder addGlobalStore(final StateStoreSupplier<KeyValueStore> storeSupplier,
-                                                      final String sourceName,
-                                                      final Deserializer keyDeserializer,
-                                                      final Deserializer valueDeserializer,
+    public synchronized StreamsBuilder addGlobalStore(final StoreBuilder storeBuilder,
```

Comment:
Hmm.. as a second thought: for this function and `addStateStore` above, should these two functions still be part of `StreamsBuilder`? I know in the old APIs they are in both DSL and PAPI, but that is mainly because the KStreamBuilder is extending TopologyBuilder, so it is easy to get the `processorName`. But now in the `StreamsBuilder` you cannot get the `processorName` any more unless you build the `Topology` and query the specific processor name.

So I'd propose an small addendum to the KIP. Here are a few thoughts I had:

1. Keep these two functions but only with one parameter `StoreBuilder`, and then in `KStream` when `process/transform` is called use `connectProcessorAndStateStores` to connect the stores.

2. Get rid of these two functions, and document that for `process/transform`, you need to call `builder.build().addStateStore()` at the end with the name that you reference inside the `Processor / Transformer`.

3. In `KStream`, deprecate `process(processorSupplier, storeNames...)` with `process(processorSupplier, StoreBuilder...)`, and remove these two functions in `StreamsBuilder`.

WDYT? @mjsax @dguy @bbejeck

Reply:
I agree on the issue. Nice catch!!

Side remark:  For `addGlobalStore()` we can keep as is, because `processorName` in not the name of an existing processor, but the name of the newly created processor that maintains the store. Thus, there is no issue with `addGlobalStore()` IMHO, as we don't need to pass in an unknown name of another processor.

Option (2) is not very user friendly IMHO and I have a strong opinion against it.

Second side remark: `connectProcessorAndStateStores()` is not available in the DSL, forcing users to call `builder.build().connectProcessorAndStateStores()`, thus, this is not an option IMHO. It's also not required as one can connect a store to a processor when adding the processor via `process(processorSupplier, storeNames...)`

If we want to stay close to old API, option (1) removing the second parameter from `addStateStore()` makes sense. The store name can be accessed easily via `StoreBuilder#name()` and thus stores can be connected via `process(processorSupplier, storeNames...)`. This is my personally favored option.

Option (3) adding `process(processorSupplier, StoreBuilder...)` might be a syntactic sugar to avoid the `addStateStore()` call. However, we would need to make sure, that two processor can be connected to the same store -- if we deprecate the `process(processorSupplier, storeNames...)` we must match stores via `StoreBuilder` (ie, object identity) -- ie, a user needs to pass in the same `StoreBuilder` to each call of `process(processorSupplier, StoreBuilder...)` to connect all processor to the same store. This, seems to be rather unintuitive and error prone to me... (eg, people try to pass in two different `StoreBuilder` that build equal stores and return the same `storeName` in order to connect two processor to one store -- but to us, it's two stores with a naming conflict and we should throw a `TopologyException` -- this might be rather confusing to users). It also moves away from PAPI usage pattern and I think we should not do that for this case.

Reply:
I agree with @mjsax - i'll update accordingly

Reply:
I'm wondering if we can actually remove `addGlobalStore` completely from `StreamsBuilder`? It is not needed by the DSL and if users where using the DSL and wanted to manually add a global store they could always use `StreamsBuilder#build` to get a `Topology` and then call `addGlobalStore` on it - WDYT?

Reply:
For option 2 seems a little awkward IMHO, I second what @mjsax says.

Reply:
I think people want to use a global store in DSL, too. And forcing people to call `build()` to add one, is not a good idea IMHO. (cf option (2)).

Also, if you argue like this, we could remove `addStore`, too, because people can also add the store via `builder.builder().addStore()` -- however, in KIP-120 discussion, it was explicitly requested to add both methods to `StreamsBuilder` to avoid this pattern.

Also note, people who are new and want to use a stateful process() will always ask: how can I add a store? There is not API on `StreamsBuilder`.

Reply:
I think in DSL, users may still wants to access a global state store in `process/transform`, that was the motivation for adding this API.

Personally I'd vote for option 2) above, to keep the APIs succinct without semi-duplicated calls. But seems you all prefer option 1) in terms of user convenience, so I'm fine with that as well.